pkg/controller/v1alpha2/experiment/experiment_controller.go

```diff
@@ -89,8 +113,127 @@ func (r *ReconcileExperiment) Reconcile(request reconcile.Request) (reconcile.Re
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
+		log.Error(err, "Experiment Get error")
 		return reconcile.Result{}, err
 	}
+	original := instance.DeepCopy()
+
+	if instance.IsCompleted() {
+
+		return reconcile.Result{}, nil
+
+	}
+	if !instance.IsCreated() {
+		//Experiment not created in DB
+		err = util.CreateExperimentInDB(instance)
+		if err != nil {
+			log.Error(err, "Create experiment in DB error")
+			return reconcile.Result{}, err
+		}
+
+		if instance.Status.StartTime == nil {
+			now := metav1.Now()
+			instance.Status.StartTime = &now
+		}
+		msg := "Experiment is created"
+		instance.MarkExperimentStatusCreated(util.ExperimentCreatedReason, msg)
+		requeue = true
+	} else {
+		// Experiment already created in DB
+		err := r.ReconcileExperiment(instance)
+		if err != nil {
+			log.Error(err, "Reconcile experiment error")
+			return reconcile.Result{}, err
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(original.Status, instance.Status) {
+		//assuming that only status change
+		err = util.UpdateExperimentStatusInDB(instance)
+		if err != nil {
+			log.Error(err, "Update experiment status in DB error")
+			return reconcile.Result{}, err
+		}
+		err = r.Status().Update(context.TODO(), instance)
+		if err != nil {
+			log.Error(err, "Update experiment instance status error")
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{Requeue: requeue}, nil
+}
+
+func (r *ReconcileExperiment) ReconcileExperiment(instance *experimentsv1alpha2.Experiment) error {
+
+	var err error
+	log := log.WithValues("Experiment", types.NamespacedName{Name: instance.GetName(), Namespace: instance.GetNamespace()})
+	trials := &trialsv1alpha2.TrialList{}
+	labels := map[string]string{"experiment": instance.Name}
+	lo := &client.ListOptions{}
+	lo.MatchingLabels(labels).InNamespace(instance.Namespace)
+
+	err = r.List(context.TODO(), lo, trials)
+	if err != nil {
+		log.Error(err, "Trial List error")
+		return err
+	}
+	if len(trials.Items) > 0 {
+		err := util.UpdateExperimentStatus(instance, trials)
+		if err != nil {
+			log.Error(err, "Update experiment status error")
+			return err
+		}
+	}
+	reconcileRequired := !instance.IsCompleted()
+	if reconcileRequired {
+		r.ReconcileTrials(instance)
+	}
+	return err
+}
+
+func (r *ReconcileExperiment) ReconcileTrials(instance *experimentsv1alpha2.Experiment) error {
+
+	var err error
+	log := log.WithValues("Experiment", types.NamespacedName{Name: instance.GetName(), Namespace: instance.GetNamespace()})
+	parallelCount := 0
+
+	if instance.Spec.ParallelTrialCount != nil {
+		parallelCount = *instance.Spec.ParallelTrialCount
+	} else {
+		parallelCount = 3
```

Comment:
I think it is useful to fillback instance.Spec.ParallelTrialCount with default value so that user can get what value it is by "kubectl get"
```
instance.Spec.ParallelTrialCount = new(int)
*instance.Spec.ParallelTrialCount = 3
```

Reply:
Agree. This can be done when validate and defaults functions are added

Reply:
We should document the default behavior in experiment_types.go

Reply:
done